File: 3.cameraandmatrices/webgl/source/js/engine/rendering/CCModelBase.js
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2011 - 2011 France Telecom
 * This software is distributed under the Apache 2.0 license.
 * http://www.apache.org/licenses/LICENSE-2.0.html
 *-----------------------------------------------------------
 * File Name   : CCModelBase.js
 * Description : Represents the attributes of a renderable object.
 *
 * Created     : 21/10/11
 * Author(s)   : Ashraf Samy Hegab
 *-----------------------------------------------------------
 */

function CCModelBase()
{
	this.construct();
}
CopyPrototype( CCModelBase, CCRenderable, "CCRenderable" );


CCModelBase.prototype.construct = function()
{
	this.CCRenderable_construct();
	
    this.primitives = new Array();
    this.models = new Array();
    
    this.colour = false;
	this.colourOutline = false;
}


CCModelBase.prototype.render = function(renderer, alpha)
{
	if( this.shouldRender )
	{
		GLPushMatrix();
		{
			this.refreshModelMatrix();
			GLMultMatrix( this.modelMatrix );
			
			var primitives = this.primitives;
			var primitivesLength = primitives.length;
			
			if( this.colourOutline != false )
			{
				renderer.CCSetColour( this.colourOutline );
				if( alpha == false && this.colourOutline.a > 0.0 )
				{
					gl.enable( gl.BLEND );
				}
				
				renderer.CCDefaultTexCoords();
				gEngine.textureManager.setTextureIndex( 0 );
				for( var i=0; i<primitivesLength; ++i )
				{
					primitives[i].renderOutline( renderer );
				}
				
				if( alpha == false && this.colourOutline.a > 0.0 )
				{
					gl.disable( gl.BLEND );
				}
			}

			if( this.colour != false )
			{
				renderer.CCSetColour( this.colour );
			}
			
			if( renderer.colour.a > 0 )
			{
				for( var i=0; i<primitivesLength; ++i )
				{
					primitives[i].render( renderer );
				}
			}

			var models = this.models;
			var length = models.length;
			for( var i=0; i<length; ++i )
			{
				models[i].render( renderer, alpha );
				if( this.colour != false )
				{
					renderer.CCSetColour( this.colour );
				}
			}
		}
		GLPopMatrix();
	}
}


CCModelBase.prototype.addModel = function(model)
{
	this.models.push( model );
}


CCModelBase.prototype.addPrimitive = function(primitive)
{
	this.primitives.push( primitive );
}


CCModelBase.prototype.setColour = function(inColour)
{
	if( this.colour == false )
	{
		this.colour = new CCColour();
	}
	this.colour.set( inColour );
}


CCModelBase.prototype.setOutline = function(inColour)
{
	if( this.colourOutline == false )
	{
		this.colourOutline = new CCColour();
	}
	this.colourOutline.set( inColour );
}

File: 4.controls/mobile/Engine/Source/Rendering/CCModelBase.cpp
/*-----------------------------------------------------------
 * 2c - Cross Platform 3D Application Framework
 *-----------------------------------------------------------
 * Copyright © 2010 – 2011 France Telecom
 * This software is distributed under the Apache 2.0 license,
 * see the "license.txt" file for more details.
 *-----------------------------------------------------------
 * File Name   : CCModelBase.cpp
 *-----------------------------------------------------------
 */

#include "CCDefines.h"
#include "CCPrimitives.h"


CCModelBase::CCModelBase()
{	
	colour = colourOutline = NULL;
    
    shader = "basic";
}


void CCModelBase::destruct()
{
	if( scale )
	{
        FREE_POINTER( scale );
	}
	
	if( colour != NULL )
	{
		delete colour;
	}
	
	if( colourOutline != NULL )
	{
		delete colourOutline;
	}
	
	models.deleteObjectsAndList();
	primitives.deleteObjectsAndList();
	
	super::destruct();
}


void CCModelBase::render(const bool alpha)
{
#if defined PROFILEON
    CCProfiler profile( "CCModelBase::render()" );
#endif

    if( shouldRender )
    {
        GLPushMatrix();
        {
            refreshModelMatrix();
            GLMultMatrixf( modelMatrix );
            
            gEngine->renderer->setShader( shader );

            if( colourOutline != NULL )
            {
#if defined PROFILEON
            	CCProfiler profile( "CCModelBase::render()::colourOutline" );
#endif
                CCSetColour( *colourOutline );
                if( alpha == false && CCColourHasAlpha() )
                {
                    glEnable( GL_BLEND );
                }

                CCDefaultTexCoords();
                gEngine->textureManager->setTextureIndex( 0 );
                for( int i=0; i<primitives.length; ++i )
                {
                    primitives.list[i]->renderOutline();
                }

                if( alpha == false && CCColourHasAlpha() )
                {
                    glDisable( GL_BLEND );
                }
            }

            if( colour != NULL )
            {
                CCSetColour( *colour );
            }

            if( alpha == false || colour == NULL || ( colour != NULL && colour->alpha > 0.0f ) )
            {
                for( int i=0; i<primitives.length; ++i )
                {
                    CCPrimitiveBase *primitive = primitives.list[i];
                    primitive->render();
                }
            }

            for( int i=0; i<models.length; ++i )
            {
                CCModelBase *model = models.list[i];
                model->render( alpha );
                if( colour != NULL )
                {
                    CCSetColour( *colour );
                }
            }
        }
        GLPopMatrix();
    }
}


void CCModelBase::addModel(CCModelBase *model)
{
    ASSERT( model != this );
	models.add( model );
}


void CCModelBase::addPrimitive(CCPrimitiveBase *primitive)
{
	primitives.add( primitive );
}


void CCModelBase::setColour(const CCColour &inColour)
{
	if( colour == NULL )
	{
		colour = new CCColour();
	}
	
	*colour = inColour;
}


void CCModelBase::setColourAlpha(const float inAlpha)
{
    if( colour == NULL )
    {
        colour = new CCColour();
    }
    colour->alpha = inAlpha;
    setColour( *colour );
}


void CCModelBase::setOutline(const CCColour &inColour)
{
	if( colourOutline == NULL )
	{
		colourOutline = new CCColour();
	}
	
	*colourOutline = inColour;
}
